### ISSUE TITLE
How to use tracked reference inside of hybrid component

### ISSUE THREAD

**Issue by rubixibuc** [hybridsjs/hybrids#104]
Sorry, two more issues incoming! I really love this library and have begun the process of converting all my components to it, so I apologize for all of the issues I'm creating

Using the following example [https://hybrids.js.org/built-in-factories/render#reference-internals](https://hybrids.js.org/built-in-factories/render#reference-internals)

If I reference the DOM element property inside of the render function itself, I get this error:

![133C5A5E-202C-470B-9DBB-C5E6C7E5E599_4_5005_c](https://user-images.githubusercontent.com/15688746/77976265-8b0bfb80-72ca-11ea-9c57-96dfb8778333.jpeg)

When using a third party component that has an anchor prop, passing the element is necessary.

example: [https://stackblitz.com/edit/hybrids-children-factory-vlnfup]( https://stackblitz.com/edit/hybrids-children-factory-vlnfup)

Is there another or better way to do this?

Ty!

**Comment by smalluban** [comment]
No problem at all! If you have any problems with the library, you are welcome to create issues - they might help others in the future ;)

I looked at your example and I am not sure what's your intention. If you are going to use `@material/mwc-menu`, there is no need to create a reference to the button from the template. You can simply attach click listener to the button, which changes an internal value passed to the menu "open" property, like this:

```javascript
function toggleMenu(host) {
  host.menuOpen = !host.menuOpen;
}

{
  menuOpen: false,
  render: ({ menuOpen }) => html`
    <button onclick=${toggleMenu}>toggle menu</button>
    <mwc-menu open=${menuOpen}>
       ...
     </mwc-menu>
  `,
}
```

One of the main ideas of the library is to use a declarative approach as much as we can. Instead of procedural action, we can control an open/close state of the menu by the property value. The action itself is a side-effect done outside of the template.

Going back to your solution, sometimes the `ref` pattern might be helpful. However, it is for building external API, to access elements from your element. In the docs example, the`canvas` element is shown. The custom element may create a DOM shell for the canvas element, but it wants to expose that element, as its API is mainly procedural (you have to take some actions with canvas element to draw). 

The error that you have seen is the correct behavior. You cannot create a circular dependency, because then you might create an endless loop - if property "a" eventually wants itself, the get call might not end.

**Comment by rubixibuc** [comment]
_mwc-menu_ has an anchor property that is used to position the menu, so I need to pass it there. I want the menu below the button.

I got this working last night by setting a property with the event.target inside the click handler. I set this as the same time as open, so they render together.

essentially:

```
const ref = (prop) => (host, event) => (host[prop] = event.target);

const Component = {
    _open: false,
    _menu: null,
    render: host => html`
        <mwc-button onclick=${_.over([html.set("_open", true), ref("_menu")])} ></mwc-button>
        <mwc-menu anchor=${host._menu} open=${host._open}></mwc-menu>
    `
}
```

If there is a better way to do this please let me know

It might be cool to add **html.ref(prop)** or **html.setRef(prop)** to the library if you want me to create a pull request.

**Comment by smalluban** [comment]
I have to say that this is a kind of use case, which the library does not support out of the box :)

Requiring JS to set anchor element in `mwc-menu` is rather odd and rare. Although, this web component also recommends using the nearest parent with `position: relative`, so it might be useful in your case as well, isn't?

> Determines from which element the floating menu should calculate sizing and position offsets. **In the default case, both mwc-menu and the anchor should share a parent with position:relative.** Changing anchor typically requires absolute or fixed.

Did you try to create a wrapper for your button and menu? It then should work the same as with your programmatic solution:

```html
<style>
  #container { position: relative }
</style>
<div id="container">
  <mwc-button></mwc-button>
  <mwc-menu></mwc-menu>
</div>
```

**Comment by rubixibuc** [comment]
I did just try that solution, and it doesn't place the menu in the correct location. Unless I pass the element I want it positioned below (the button), it places it on top of the button, not below.

I was able to generate a reference to the element using the code in my comment above. So I should be all set even without having a generally supported way to do this. I was more curious if there was a built in way.

Ty :)
